Medical QoL Update

Dave-TH opened the pull request:
I've got a few nifty little updates that will hopefully make medical a bit more fun!

IV drips have been completely refactored. There isn't a ton of change on the frontend, but they've got nifty new sprites courtesy of Goonstation You may now put any chemical directly into the IV bag. Medical starts with a few saline-glucose bags, and chemistry has a box of empty spares for filling with whatever the doctors might need!

I've also ported /tg/'s [wall-mounted defib units](https://github.com/tgstation/tgstation/pull/34545).
These frames:
 - Mount to the wall and allow anyone to hold the paddles.
 - Charge the inserted defib unit from the powernet.
 - Lock with a (Head/Medical/Security) cardswipe to prevent theft.

Medical starts with one in the exam center right next to the sleepers, and R&D can print more for a low cost.

![wehopetomakedeathmoremeaningful](https://user-images.githubusercontent.com/16298462/52834834-b8d7b600-30b1-11e9-90ea-eede2073bc11.PNG)

How medical looks now:
![chompchomp](https://user-images.githubusercontent.com/16298462/52887738-dd31a200-3146-11e9-9e8b-6dede956af19.PNG)


:cl:
add: Wall-mounted defib frames. Swipe an appropriate card to lock the defib in place and use it to save lives!
tweak: IV bags have been refactored and may now accept any chemical. Play around with them a bit!
/:cl:

Aurorablade commented:
So how many tries it t take to get the screenshot right?

Fox-McCloud commented on code/modules/reagents/reagent_containers/iv_bag.dm:
```diff
@@ -0,0 +1,182 @@
+	update_icon()
+
+/obj/item/reagent_containers/iv_bag/proc/begin_processing(mob/target)
+	to_chat(world, "DEBUG: [src] processing started with target [target].")
```
👀

Dave-TH replied:
Feature™

Dave-TH commented:
Amended debug statements left in totally on purpose, fixed outstanding /blood/empty subtypes that no longer exist.

Dave-TH commented:
IV drips now make a soft rolling sound when moved

EvadableMoxie commented:
I like the idea of wall mounted defibs, but I wouldn't want one _there._  It's far from anything you'd want to have nearby when you revive someone (namely cryo), and it's at the end of a 1 tile choke of hallway. That's going to be an utter nightmare to use in any high traffic situation. I'd probably stick with the one near cryo than bother with that one in most cases.

AffectedArc07 commented:
+1 for the IV refactor
+1 for the defib holder
+1 for the massacred vulp

This should make medbay more impactful

Mitchs98 commented:
I like it. But agree on placement. Should probably be near cryo so it's more useful.

Ralta commented:
Yeeeeeeees. But an alternate suggestion: since the defib frame can be printed, perhaps give the initial one to the CMO instead of having it pre-placed? That gives the CMO a choice on where to place it. It also means a fast antag could steal it for _insert nefarious purpose here_.

Mitchs98 commented:
Did not read the part about being able to print it...I'm dumb. Haha...

MrMagolor commented:
The R&D wall-mounted defibs should have the same tech costs as normal defibs, no?

Dave-TH commented:
@Ralta The frame can be unwrenched if you yank the defib out. This means it can be relocated, but if the defib is locked in there with a card swipe you can't use a wrench to force it out.

@MrMagolor It's just the frame. You still have to put a defib inside of it.

Dave-TH commented:
Also, somewhat important bit I forgot: The rules for injecting with an IV bag are the same as injecting with a syringe. Thick materials will block it!

UmeFuu commented:
As a main med, I totally support this.
The change of sprite isn't my favourite, but I guess It's just different.
I can't wait to inyect 3 different IV drips with water, salt and sugar to the same person and similar mixes just for the sake of it.

davidchan13 commented:
Why does Security get access?

KasparoVy commented:
IV cocaine
RIP vulp, you gave your fuzzy life for medical science

Dave-TH commented:
Minor tweaks. Added a few comments, fixed using reagent containers on IV stands to add chems to the attached bag, removed sec access from the defib mount.

Updated the OP with an image of where the defib is actually placed.

Mitchs98 commented:
Just to clarify; Medbay still starts with bloodbags, yes? The IV bags are only an addition? It's probably easy to assume they do, but, you know.

Fox-McCloud commented on code/game/machinery/defib_mount.dm:
```diff
@@ -0,0 +1,156 @@
+
+/obj/machinery/defibrillator_mount/Destroy()
+	if(defib)
+		QDEL_NULL(defib)
```
`QDEL_NULL` already has safeties build into it, so you don't have to check for the defib first.

Fox-McCloud commented on code/game/machinery/defib_mount.dm:
```diff
@@ -0,0 +1,156 @@
+		update_icon()
+
+/obj/machinery/defibrillator_mount/update_icon()
+	cut_overlays()
```
I'm not sure how well the `cut_overlays` and `add_overlays` system works with objects. I know it was implemented for mobs.

Just be doubly sure this doesn't cause any oddities or runtimes.

Dave-TH replied:
This is how syringes and similar reagent containers handle it AFAIK.

Fox-McCloud replied:
they use `overlays.Cut()` and `overlays += "thing"`

`cut_overlays` is a TG proc with their icon SS. as is `add_overlay`

taukausanake replied:
I think I messed with `cut_overlays()` in one of my little side projects and it didn't work until someone changed it to `overlays.cut()`. I think Fox is correct

Dave-TH replied:
These procs are implemented at the atom level and work as expected in this case.

Fox-McCloud commented on code/modules/reagents/reagent_containers/iv_bag.dm:
```diff
@@ -0,0 +1,180 @@
+		if(reagents.total_volume < reagents.maximum_volume)
+			injection_target.transfer_blood_to(src, amount_per_transfer_from_this)
+			for(var/datum/reagent/x in injection_target.reagents.reagent_list) // Pull small amounts of reagents from the person while drawing blood
+				injection_target.reagents.trans_to(src, amount_per_transfer_from_this/10)
```
This is probably about the only thing I don't like; it's a mini-dialysis machine, albeit, not a particularly good one.

Dave-TH replied:
With how little sleeper dialysis is used I'm not sure I'm super worried about that as a balance point. The small amount of chems it pulls out are more for flavor/consistency than and mechanical benefit.

farie82 commented:
One issue I think of instantly.  An IV bag is attached instantly. No warning either. This can be used to cheese the hell out of meme chems or such. Unless there is an attach time and a warning

Fox-McCloud commented:
come on Travis, don't be a butt.

Dave-TH commented:
@farie82 Exact same rules for using a syringe. Instant on yourself, same timer for others assuming thickmaterial doesn't block it.

@Mitchs98 No, I did not remove all the blood from medical.

Fox-McCloud commented on code/game/machinery/defib_mount.dm:
```diff
@@ -0,0 +1,155 @@
+	var/clamps_locked = FALSE //if true, and a defib is loaded, it can't be removed without unlocking the clamps
+
+/obj/machinery/defibrillator_mount/New(location, direction, building = 0)
+	. = ..()
```
Don't think we need to return the parent value.

Fox-McCloud commented on code/game/machinery/defib_mount.dm:
```diff
@@ -0,0 +1,155 @@
+
+
+/obj/machinery/defibrillator_mount/loaded/New() //loaded subtype for mapping use
+	. = ..()
```
same

Fox-McCloud commented on code/game/machinery/defib_mount.dm:
```diff
@@ -0,0 +1,155 @@
+
+/obj/machinery/defibrillator_mount/Destroy()
+	QDEL_NULL(defib)
+	. = ..()
```
purely preference, though this is technically fine, but I always preferred to use `return ..()` if it's at the end; save `. = ..()` for when you're not using it at the end.

Fox-McCloud commented on code/game/machinery/defib_mount.dm:
```diff
@@ -0,0 +1,155 @@
+		else
+			to_chat(user, "<span class='warning'>Insufficient access.</span>")
+		return
+	..()
```
return the parent value here.

Fox-McCloud commented on code/game/machinery/defib_mount.dm:
```diff
@@ -0,0 +1,155 @@
+/obj/machinery/defibrillator_mount/AltClick(mob/living/carbon/user)
+	if(!istype(user))
+		return
+	if(user.stat || user.restrained())
```
`incapacitated`

Fox-McCloud commented on code/game/machinery/iv_drip.dm:
```diff
@@ -71,167 +26,45 @@
-			update_icon()
-		else
-			to_chat(usr, "<span class='warning'>There's nothing attached to the IV drip!</span>")
+		bag.afterattack(target, usr, 1)
```
`TRUE` on the proximity argument.

Fox-McCloud commented on code/game/machinery/iv_drip.dm:
```diff
@@ -71,167 +26,45 @@
+/obj/machinery/iv_drip/attack_hand(mob/user)
+	if(bag)
+		user.put_in_hands(bag)
+	bag.update_icon()
```
if there's no bag, this is going to runtime.

Were the last 3 lines of this proc supposed to be indented one more?

Fox-McCloud commented on code/game/machinery/iv_drip.dm:
```diff
@@ -71,167 +26,45 @@
-		attached = null
+	else if (bag && istype(I, /obj/item/reagent_containers))
+		bag.attackby(I)
+		I.afterattack(bag, usr, 1)
```
`TRUE`

Fox-McCloud commented on code/game/machinery/iv_drip.dm:
```diff
@@ -71,167 +26,45 @@
+	. = ..()
+	if(!.) // ..() will return 0 if we didn't actually move anywhere.
+		return .
+	playsound(loc, pick('sound/items/cartwheel1.ogg', 'sound/items/cartwheel2.ogg'), 100, 1, ignore_walls = FALSE)
```
oh boy.

Fox-McCloud commented on code/game/objects/items/weapons/storage/boxes.dm:
```diff
@@ -191,6 +191,21 @@
+	desc = "A box full of empty IV bags."
+	icon_state = "beaker"
+
+	New()
```
I realize the rest of the file is gross, but at least make this one fine in terms of absolute pathing.

Fox-McCloud commented on code/modules/reagents/reagent_containers/iv_bag.dm:
```diff
@@ -0,0 +1,180 @@
+	if(!target.reagents)
+		return
+
+	var/mob/living/L
```
since this is only ever used after the `isliving` check, it should go under there.

Fox-McCloud commented on code/modules/reagents/reagent_containers/iv_bag.dm:
```diff
@@ -0,0 +1,180 @@
+	name = "[initial(name)] ([new_label])"
+
+/obj/item/reagent_containers/iv_bag/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/pen) || istype(I, /obj/item/flashlight/pen))
```
one day we should really probably implement a `is_writing_implement` helper. No I don't expect you to do this, just a mental note for myself.

Fox-McCloud commented on code/modules/reagents/reagent_containers/iv_bag.dm:
```diff
@@ -0,0 +1,180 @@
+/obj/item/reagent_containers/iv_bag/salglu/New()
+	..()
+	set_label("Saline-Glucose")
+	reagents.add_reagent("salglu_solution", 200)
```
Use `list_reagents`, it's much cleaner.

Fox-McCloud commented on code/modules/reagents/reagent_containers/iv_bag.dm:
```diff
@@ -0,0 +1,180 @@
+	if(blood_type != null)
+		set_label(blood_type)
+		reagents.add_reagent("blood", 200, list("donor"=null,"viruses"=null,"blood_DNA"=null,"blood_type"=blood_type,"resistances"=null,"trace_chem"=null))
+		amount_per_transfer_from_this = 5 // Bloodbags are set to transfer 5 units by default.
```
set this on the object itself, rather than in `New`

Fox-McCloud commented:
Just a few things.

Fox-McCloud commented on code/modules/reagents/reagent_containers/iv_bag.dm:
```diff
@@ -0,0 +1,179 @@
+/obj/item/reagent_containers/iv_bag/salglu/New()
+	..()
+	set_label("Saline-Glucose")
+	reagents = list_reagents = list("salglu_solution" = 200)
```
this isn't how this works; this needs to be on the defined type; using it after `New` is not going to do anything.

Fox-McCloud commented on code/game/objects/items/weapons/storage/boxes.dm:
```diff
@@ -191,6 +191,21 @@
+	icon_state = "beaker"
+
+/obj/item/storage/box/iv_bags/New()
+		..()
```
Uhhhh, two tabs?

Fox-McCloud commented on code/game/objects/items/weapons/storage/boxes.dm:
```diff
@@ -191,6 +191,21 @@
+
+/obj/item/storage/box/iv_bags/New()
+		..()
+		new /obj/item/reagent_containers/iv_bag( src )
```
`( src )`

Ew.

Fox-McCloud commented on code/modules/reagents/reagent_containers/iv_bag.dm:
```diff
@@ -0,0 +1,179 @@
+		var/image/filling = image('icons/goonstation/objects/iv.dmi', src, "[icon_state][percent]")
+
+		filling.icon += mix_color_from_reagents(reagents.reagent_list)
+		overlays += filling
```
You used `cut_overlays` and `add_overlay` for the defib mount, why not this?

Dave-TH replied:
Because the defib mount is a /tg/ port and the procs never reported any errors. Overlays behaved as expected. The bags are my original code.

Fox-McCloud commented on code/game/machinery/iv_drip.dm:
```diff
@@ -3,65 +3,20 @@
+		if(bag.reagents.total_volume)
+			var/image/filling = image('icons/goonstation/objects/iv.dmi', src, "hangingbag-fluid")
+			filling.icon += mix_color_from_reagents(bag.reagents.reagent_list)
 			overlays += filling
```
`add_overlay`

Fox-McCloud commented on code/game/machinery/iv_drip.dm:
```diff
@@ -3,65 +3,20 @@
-		else
-			icon_state = "donateidle"
-
 	overlays.Cut()
```
`cut_overlays`

variableundefined commented:
This needs an update

Aurorablade commented:
Its only been 5 days give him some time.

variableundefined commented:
It's conflicted at the moment

variableundefined commented:
Inform me when deconflicted

Dave-TH commented:
Deconflicted
